llms/mlx_lm/sample_utils.py

```diff
@@ -0,0 +1,38 @@
+import mlx.core as mx
+
+
+def top_p_filtering(logits: mx.array, top_p: float, temperature: float) -> mx.array:
+    """
+    Apply top-p (nucleus) filtering to logits using cumulative probabilities.
+
+    Args:
+        logits: The logits from the model's output.
+        top_p: The cumulative probability threshold for top-p filtering.
+        temperature: Temperature parameter for softmax distribution reshaping.
+    Returns:
+        token selected based on the top-p criterion.
+    """
+    if (
+        logits.dtype == mx.bfloat16
+    ):  # workaround for unable to load kernel contiguous_scan_inclusive_sum_bfloat16_bfloat16
+        logits = logits.astype(mx.float32)
+
+    # referenced implementation from https://github.com/huggingface/transformers/blob/main/src/transformers/generation/logits_process.py#L449-L460
+    probs = mx.softmax(logits / temperature, axis=-1)
+
+    # sort probs in ascending order
+    sorted_probs = mx.sort(probs, axis=-1)
```

Comment:
Why not `sorted_probs = probs[..., sorted_indices]`. It may be faster. Obviously having a multi-output primitive that returns both would be the best here.